src/foam/nanos/ruler/RuleEngine.java

```diff
@@ -161,11 +163,17 @@ private void asyncApplyRules(List<Rule> rules, FObject obj, FObject oldObj) {
         if ( rule.getAsyncAction() != null
           && rule.f(x, obj, oldObj)
         ) {
+          // We assume the original object `obj` is stale when running after rules.
+          // For that, greedy mode is used for object reload. For before rules,
+          // object reload uses non-greedy mode so that changes on the original
+          // object will be copied over to the reloaded object.
+          FObject nu = getDelegate().find_(x, obj);
+          reloadObject(obj, oldObj, nu, rule.getAfter());
```

Comment:
argument 'nu' is potentially reassigned in reloadObject, why does reloadObject not return the appropriate object?

Reply:
Fixed.